Godot don't compile on Linux

Can't compile 1bc91848e3cee91eccaf2150a74deaf1cd84be13 build on Linux

**If you have this problem the solution is in this comment:** https://github.com/godotengine/godot/issues/3111#issuecomment-166384446

**_Current log here (after discussion and installing libevdev-dev):_**
https://github.com/godotengine/godot/issues/3111#issuecomment-166375495
```
kubeczek@kubeczek-K52JT:~/programs/godot/20151221/godot-master$ scons platform=x11 target=release_debug -j 4
scons: Reading SConscript files ...
sh: 1: i586-mingw32msvc-gcc: not found
sh: 1: i686-w64-mingw32-gcc: not found
PulseAudio development libraries not found, disabling driver
scons: done reading SConscript files.
scons: Building targets ...
g++ -o platform/x11/joystick_linux.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Iplatform/x11/vorbis platform/x11/joystick_linux.cpp
platform/x11/joystick_linux.cpp:37:21: fatal error: libudev.h: No such file or directory
compilation terminated.
g++ -o modules/gridmap/register_types.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Imodules/gridmap/vorbis modules/gridmap/register_types.cpp
g++ -o modules/gridmap/grid_map.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Imodules/gridmap/vorbis modules/gridmap/grid_map.cpp
scons: *** [platform/x11/joystick_linux.x11.opt.tools.64.o] Error 1
g++ -o modules/gdscript/gd_editor.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Imodules/gdscript/vorbis modules/gdscript/gd_editor.cpp
scons: building terminated because of errors.
```

**obsolete log: (without libevdev-dev package), don't pay attention to this one**
```
kubeczek@kubeczek-K52JT:~/programs/godot/20151221/godot-master$ scons platform=x11 target=release_debug -j 4
scons: Reading SConscript files ...
Package libevdev was not found in the pkg-config search path.
Perhaps you should add the directory containing `libevdev.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libevdev' found
evdev not found.. x11 disabled.
sh: 1: i586-mingw32msvc-gcc: not found
sh: 1: i686-w64-mingw32-gcc: not found
No valid target platform selected.
The following were detected:
    haiku
    server
    windows
 
Please run scons again with argument: platform=<string>
scons: done reading SConscript files.
scons: Building targets ...
scons: `.' is up to date.
scons: done building targets.
```

Hmm actually it still don't compile in the end  on Ubuntu. The output is different however: 
```
kubeczek@kubeczek-K52JT:~/programs/godot/20151221/godot-master$ scons platform=x11 target=release_debug -j 4
scons: Reading SConscript files ...
sh: 1: i586-mingw32msvc-gcc: not found
sh: 1: i686-w64-mingw32-gcc: not found
PulseAudio development libraries not found, disabling driver
scons: done reading SConscript files.
scons: Building targets ...
g++ -o platform/x11/joystick_linux.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Iplatform/x11/vorbis platform/x11/joystick_linux.cpp
platform/x11/joystick_linux.cpp:37:21: fatal error: libudev.h: No such file or directory
compilation terminated.
g++ -o modules/gridmap/register_types.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Imodules/gridmap/vorbis modules/gridmap/register_types.cpp
g++ -o modules/gridmap/grid_map.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Imodules/gridmap/vorbis modules/gridmap/grid_map.cpp
scons: *** [platform/x11/joystick_linux.x11.opt.tools.64.o] Error 1
g++ -o modules/gdscript/gd_editor.x11.opt.tools.64.o -c -O2 -ffast-math -DDEBUG_ENABLED -DFREETYPE_ENABLED -DOC_X86_ASM -DOPENGL_ENABLED -DGLEW_ENABLED -DALSA_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DGLES2_ENABLED -DGLES_OVER_GL -DMUSEPACK_ENABLED -DOPENSSL_ENABLED -DSQUISH_ENABLED -DVORBIS_ENABLED -DOPUS_ENABLED -DTHEORA_ENABLED -DTHEORALIB_ENABLED -DPNG_ENABLED -DDDS_ENABLED -DPVR_ENABLED -DJPG_ENABLED -DWEBP_ENABLED -DSPEEX_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DXML_ENABLED -DETC1_ENABLED -Icore -Icore/math -Itools -Idrivers -I. -Iplatform/x11 -I/usr/include/libevdev-1.0 -I/usr/include/freetype2 -Imodules/gdscript/vorbis modules/gdscript/gd_editor.cpp
scons: building terminated because of errors.

```

--- a/platform/x11/detect.py
+++ b/platform/x11/detect.py
@@ -45,10 +45,6 @@ def can_build():
 		print("xinerama not found.. x11 disabled.")
 		return False
 
-	x11_error=os.system("pkg-config libevdev --modversion > /dev/null ")
-	if (x11_error):
-		print("evdev not found.. x11 disabled.")
-		return False
 
 	return True # X11 enabled
   
@@ -133,7 +129,6 @@ def configure(env):
 	env.ParseConfig('pkg-config x11 --cflags --libs')
 	env.ParseConfig('pkg-config xinerama --cflags --libs')
 	env.ParseConfig('pkg-config xcursor --cflags --libs')
-	env.ParseConfig('pkg-config libevdev --cflags --libs')
 
 	if (env["openssl"]=="yes"):
 		env.ParseConfig('pkg-config openssl --cflags --libs')
@@ -155,7 +150,18 @@ def configure(env):
 	env.Append(CPPFLAGS=['-DOPENGL_ENABLED','-DGLEW_ENABLED'])
 	if platform.system() == 'Linux':
 		env.Append(CPPFLAGS=["-DALSA_ENABLED"])
-		env.Append(LIBS=['asound', 'udev'])
+		env.Append(LIBS=['asound'])
+
+		if not os.system("pkg-config --exists libudev"):
+			if not os.system("pkg-config --exists libevdev"):
+				print("Enabling udev/evdev")
+				env.Append(CPPFLAGS=["-DJOYDEV_ENABLED"])
+				env.ParseConfig('pkg-config libudev --cflags --libs')
+				env.ParseConfig('pkg-config libevdev --cflags --libs')
+			else:
+				print("libevdev development libraries not found, disabling gamepad support")
+		else:
+			print("libudev development libraries not found, disabling gamepad support")
 
 	if (env["pulseaudio"]=="yes"):
 		if not os.system("pkg-config --exists libpulse-simple"):

--- a/platform/x11/joystick_linux.cpp
+++ b/platform/x11/joystick_linux.cpp
@@ -28,7 +28,7 @@
 /*************************************************************************/
 
 //author: Andreas Haas <hondres,  liugam3@gmail.com>
-#ifdef __linux__
+#ifdef JOYDEV_ENABLED
 
 #include "joystick_linux.h"
 #include "print_string.h"

--- a/platform/x11/joystick_linux.h
+++ b/platform/x11/joystick_linux.h
@@ -30,7 +30,7 @@
 //author: Andreas Haas <hondres,  liugam3@gmail.com>
 #ifndef JOYSTICK_LINUX_H
 #define JOYSTICK_LINUX_H
-#ifdef __linux__
+#ifdef JOYDEV_ENABLED
 #include "main/input_default.h"
 #include "os/thread.h"
 #include "os/mutex.h"

--- a/platform/x11/os_x11.cpp
+++ b/platform/x11/os_x11.cpp
@@ -426,7 +426,7 @@ void OS_X11::initialize(const VideoMode& p_desired,int p_video_driver,int p_audi
 	physics_2d_server->init();
 
 	input = memnew( InputDefault );
-#ifdef __linux__
+#ifdef JOYDEV_ENABLED
 	joystick = memnew( joystick_linux(input));
 #endif
 	_ensure_data_dir();
@@ -461,7 +461,7 @@ void OS_X11::finalize() {
 
 	physics_2d_server->finish();
 	memdelete(physics_2d_server);
-#ifdef __linux__
+#ifdef JOYDEV_ENABLED
 	memdelete(joystick);
 #endif
 	memdelete(input);
@@ -1753,7 +1753,7 @@ void OS_X11::run() {
 	while (!force_quit) {
 	
 		process_xevents(); // get rid of pending events
-#ifdef __linux__
+#ifdef JOYDEV_ENABLED
 		event_id = joystick->process_joysticks(event_id);
 #endif
 		if (Main::iteration()==true)

--- a/platform/x11/os_x11.h
+++ b/platform/x11/os_x11.h
@@ -126,7 +126,7 @@ class OS_X11 : public OS_Unix {
 
 	InputDefault *input;
 
-#ifdef __linux__
+#ifdef JOYDEV_ENABLED
 	joystick_linux *joystick;
 #endif
 

